### PR TITLE
[server] Every time we stop a workspace instance, log the reason why

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Every time we stop a workspace instance, log the reason why. This should help with debugging.

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/12283

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

File: components/server/src/user/user-deletion-service.ts [hunk 1]
```diff
@@ -113,6 +113,12 @@ export class UserDeletionService {
             runningWorkspaces.map(async (info) => {
                 const wsi = info.latestInstance;
 
+                const reason = "deleting user";
+                log.info(
+                    { workspaceId: info.workspace.id, instanceId: wsi.id },
+                    `Stopping workspace instance (reason: ${reason})`,
```
Comment (geropl) [inline components/server/src/user/user-deletion-service.ts:119]: :bulb: Could also be a payload, so it's easier to write log queries for it:

`..., "Stopping workspace instance", { reason });`

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 2]
```diff
@@ -748,14 +748,15 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
...
-        this.internalStopWorkspace(ctx, workspace).catch((err) => {
+        this.internalStopWorkspace(ctx, workspace, "stopped via API").catch((err) => {
             log.error(logCtx, "stopWorkspace error: ", err);
         });
     }
 
     protected async internalStopWorkspace(
         ctx: TraceContext,
         workspace: Workspace,
+        reason: string,
```
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:759]: :cloud: Maybe this should be optional? :thinking:

File: components/server/src/workspace/workspace-starter.ts [hunk 3]
```diff
@@ -439,8 +439,12 @@ export class WorkspaceStarter {
         ctx: TraceContext,
         instanceId: string,
         instanceRegion: string,
+        reason: string,
```
Comment (geropl) [inline components/server/src/workspace/workspace-starter.ts:442]: same as above: maybe optional? :thinking:

File: components/server/src/workspace/workspace-starter.ts [hunk 4]
```diff
@@ -439,8 +439,12 @@ export class WorkspaceStarter {
...
         policy?: StopWorkspacePolicy,
     ): Promise<void> {
+        ctx.span?.setTag("stopWorkspaceReason", reason);
+        log.info({ instanceId }, `Stopping workspace instance (reason: ${reason})`);
```
Comment (geropl) [inline components/server/src/workspace/workspace-starter.ts:446]: same as [above](https://github.com/gitpod-io/gitpod/pull/12963/files#r971722712).

**Review by geropl — APPROVED** [review]
Thx @jankeromnes , LGTM.

/hold In case you want to tackle the nits, and to give @sagor999 the possibility to review as well.

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 2 above)
Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:759]: Why make it optional if we always want to know the reason a workspace was stopped?

**Review by sagor999 — APPROVED** [review]
:heart:

**Comment by jankeromnes** [comment]
Thanks for the reviews!

Nits addressed, go for launch 🚀

/unhold

**Pull request merged**